### DISCUSSION TITLE
It seems can't catch Exception "too many file descriptors in select()"

### DISCUSSION THREAD

**Discussion by binaryantcom** [fastapi/fastapi#9399] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import uvicorn
from fastapi import FastAPI
from fastapi.responses import PlainTextResponse
import aiomysql
import asyncio
import  sys
from fastapi import FastAPI, HTTPException
from fastapi.exceptions import RequestValidationError
from fastapi.responses import PlainTextResponse
from starlette.exceptions import HTTPException as StarletteHTTPException

app = FastAPI()


if sys.platform == 'win32':
    loop = asyncio.ProactorEventLoop()
    asyncio.set_event_loop(loop)

# @app.get("/")
# async def root():
#     # Create a database instance, and connect to it.
#     await database.connect()
#     query = "select * from data limit 100,100"
#     rows = await database.fetch_all(query=query)
#     await database.disconnect()

#     return {"message": "Hello World"}
 
@app.exception_handler(StarletteHTTPException)
async def http_exception_handler(request, exc):
    return PlainTextResponse(str(exc.detail), status_code=exc.status_code)


@app.exception_handler(RequestValidationError)
async def validation_exception_handler(request, exc):
    return PlainTextResponse(str(exc), status_code=400)

 

@app.exception_handler(Exception)
async def exception_handler(request, exc):
    print('-------------')
    print(type(exc))
    print(exc)
    print('------------------')
    return PlainTextResponse("a")


@app.get("/")
async def root(): 
    pool = await aiomysql.create_pool(host="127.0.0.1", port=3306,
                                user='root', password='wnhf888cx',
                                db='mysql', charset='utf8')
    async with pool.acquire() as conn:
        async with conn.cursor() as cur:
            await cur.execute("select * from help_topic limit 100,100")
            rows=await cur.fetchall()
            # print(rows)
        
    pool.close()
    await pool.wait_closed()
  
    # Create a database instance, and connect to it.
    # conn = await aiomysql.connect(host="127.0.0.1", port=3306,
    #                               user='root', password='123456',
    #                               db='mysql', charset='utf8')
    # cursor = await conn.cursor()
    # await cursor.execute('select * from help_topic limit 100,100')
    # result = await cursor.fetchall()

    # await cursor.close()
    # conn.close()

    return {"message": "Hello World"}
 

if __name__ == '__main__':
    uvicorn.run(app="s:app", host="localhost", port=7000,log_level='error',workers=4)
```

### Description

@app.exception_handler(StarletteHTTPException)
async def http_exception_handler(request, exc):
    return PlainTextResponse(str(exc.detail), status_code=exc.status_code)

@app.exception_handler(RequestValidationError)
async def validation_exception_handler(request, exc):
    return PlainTextResponse(str(exc), status_code=400)

 

@app.exception_handler(Exception)
async def exception_handler(request, exc):
    print('-------------')
    print(type(exc))
    print(exc)
    print('------------------')
    return PlainTextResponse("a")

**The three functions above never in...**
I use ab test ab -n 10000 -c 1000 http://localhost:7000/,I don't want to miss every exception

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "D:\python\lib\multiprocessing\process.py", line 314, in _bootstrap
    self.run()
  File "D:\python\lib\multiprocessing\process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "D:\python\lib\site-packages\uvicorn\_subprocess.py", line 76, in subprocess_started
    target(sockets=sockets)
  File "D:\python\lib\site-packages\uvicorn\server.py", line 59, in run
    return asyncio.run(self.serve(sockets=sockets))
  File "D:\python\lib\asyncio\runners.py", line 47, in run
    _cancel_all_tasks(loop)
  File "D:\python\lib\asyncio\runners.py", line 63, in _cancel_all_tasks
    loop.run_until_complete(tasks.gather(*to_cancel, return_exceptions=True))   
  File "D:\python\lib\asyncio\base_events.py", line 636, in run_until_complete
    self.run_forever()
  File "D:\python\lib\asyncio\base_events.py", line 603, in run_forever
    self._run_once()
  File "D:\python\lib\asyncio\base_events.py", line 1868, in _run_once
    event_list = self._selector.select(timeout)
  File "D:\python\lib\selectors.py", line 324, in select
    r, w, _ = self._select(self._readers, self._writers, [], timeout)
  File "D:\python\lib\selectors.py", line 315, in _select
    r, w, x = select.select(r, w, w, timeout)
ValueError: too many file descriptors in select()

### Operating System

Windows

### FastAPI Version

0.95.0

### Python Version

3.10.9

**Comment by jgould22** [comment, marked as answer]
I do not use windows so take my suggestions with a grain of salt.

You are creating a connection pool in every request (generally you do not want to do this), which means you are opening a connection to the database in every request.

That error seems to be related to windows limiting the number of connections (1000 when you use that ab command).

In order to avoid that you have that bit at the top to set the loop to the the proactor event loop 
```python 
if sys.platform == 'win32':
    loop = asyncio.ProactorEventLoop()
    asyncio.set_event_loop(loop)
```

However it appears when you call `uvicorn.run` it is setting the loop policy to `WindowsSelectorEventLoop` https://github.com/encode/uvicorn/blob/411dd87ed0574a0581d2b7e5ff79de74ed3c9868/uvicorn/loops/asyncio.py#L10

I am unsure how you go about changing that, maybe it is an option in the uvicorn config object.

Another option may be to share the connection pool between requests (this is the better architecture)  and not create a new one for every request.

Best of luck in your debugging.
